### ISSUE TITLE
CIGAR alignment length changed for sequence "*"

### ISSUE THREAD

**Issue by jkbonfield** [enasequence/cramtools#35]
A SAM sequence of "*" with a CIGAR aliginment of 10M gets the alignment converted to 0M after passing through cramtools 3.0 (2.1 blows up).  Additionally it computes the checksum differently to htslib, but it is unclear which is incorrect.  In part parts:

Part 1)   Using htslib/test/c1#noseq.sam
```
jkb@seq3a[htslib/test] /software/bin/java -Xmx4000m -jar /nfs/users/nfs_j/jkb/work/cram/cramtools/cramtools-3.0.jar cram -R c1.fa -I c1#noseq.sam -O _tmp.cram -n -Q --capture-all-tags
jkb@seq3a[htslib/test] /software/bin/java -Xmx4000m -jar /nfs/users/nfs_j/jkb/work/cram/cramtools/cramtools-3.0.jar bam -R c1.fa -I _tmp.cram -O _tmp.sam
jkb@seq3a[htslib/test] grep -n SQ1 c1#noseq.sam _tmp.sam
c1#noseq.sam:4:SQ1      0       c1      1       0       10M     *       0       0       *       *  MD:Z:10  NM:i:0
_tmp.sam:7:SQ1  0       c1      1       0       0M      *       0       0       *       *
```

It appears to be because it's stored read length 0 in the file, which is the length of the BAM sequence, but not the length that is needed in order to reconstruct the CIGAR string.  Cram_dump shows:

Rec 3/9 at 1,3
BF = 0 => SAM 0x0 (ret 0, out_sz 1)
CF = 10 (ret 0, out_sz 1)
RL = 0 (ret 0, out_sz 1)
AP = 0 (ret 0, out_sz 1)
RG = -1 (ret 0, out_sz 1)
RN = SQ1 (ret 0, out_sz 3)
Detached
MF = 0 (ret 0, out_sz 1)
NS = -1 (ret 0, out_sz 1)
NP = 0 (ret 0, out_sz 1)
TS = 0 (ret 0, out_sz 1)
TL = 0 (ret 0, out_sz 1)
FN = 0 (ret 0, out_sz 1)
MQ = 0 (ret 0, out_sz 1)

If I create the cram file using htslib then Cramtools can decode it and get the 10M back again.  Cram_dump shows the difference:

Rec 3/9 at 1,3
BF = 0 => SAM 0x0 (ret 0, out_sz 1)
CF = 11 (ret 0, out_sz 1)
RL = 10 (ret 0, out_sz 1)
AP = 0 (ret 0, out_sz 1)
RG = -1 (ret 0, out_sz 1)
RN = SQ1 (ret 0, out_sz 3)
Detached
MF = 0 (ret 0, out_sz 1)
NS = -1 (ret 0, out_sz 1)
NP = 0 (ret 0, out_sz 1)
TS = 0 (ret 0, out_sz 1)
TL = 1 (ret 0, out_sz 1)
  0: TN= MDZ
id=5063770
  0: Val 31 30 00
  1: TN= NMC
id=5131587
  1: Val 00
FN = 0 (ret 0, out_sz 1)
MQ = 0 (ret 0, out_sz 1)
QS =            (ret 0, out_sz 10)

Note here it has also stored the MD:Z: and NM:i tag verbatim as it cannot be sure that they would be recreated correctly during decoding as it has no sequence to work from.

Part 2)

Decoding the above htslib generated file, cramtools-3.0.jar gives me an error regarding the sequence checksums:

jkb@seq3a[htslib/test] /software/bin/java -Xmx4000m -jar /nfs/users/nfs_j/jkb/work/cram/cramtools/cramtools-3.0.jar bam -R c1.fa -I _tmp.cram -O _tmp.sam
ERROR   2015-04-16 15:07:16     ContentDigests  Content hash mismatch for tag SD, actual: 06607F2A; expected: 8244A410

The actual sequence shown doesn't seem to differ between the input and output files:

```
jkb@seq3a[htslib/test] cat c1#noseq.sam _tmp.sam
@SQ     SN:c1   LN:10
sq1     0       c1      1       0       10M     *       0       0       AACCGCGGTT      ********** MD:Z:10  NM:i:0
sQ1     0       c1      1       0       10M     *       0       0       AACCGCGGTT      *       MD:Z:10     NM:i:0
SQ1     0       c1      1       0       10M     *       0       0       *       *       MD:Z:10 NM:i:0
sq2     0       c1      1       0       4M1D5M  *       0       0       AACCCGGTT       *********  MD:Z:4^G5        NM:i:1
sQ2     0       c1      1       0       4M1D5M  *       0       0       AACCCGGTT       *       MD:Z:4^G5   NM:i:1
SQ2     0       c1      1       0       4M1D5M  *       0       0       *       *       MD:Z:4^G5  NM:i:1
sq3     4       c1      1       0       *       *       0       0       AACCCGGTT       *********
sQ3     4       c1      1       0       *       *       0       0       AACCCGGTT       *
SQ3     4       c1      1       0       *       *       0       0       *       *
@HD     VN:1.4  SO:unsorted
@SQ     SN:c1   LN:10   UR:/nfs/users/nfs_j/jkb/work/samtools_master/htslib/test/c1.fa  M5:e224bd8a70e5466c6a033be886045c85
@PG     ID:0    PN:cramtools    VN:3.0-b12      CL:java /nfs/users/nfs_j/jkb/work/cram/cramtools/cramtools-3.0.jar bam -R c1.fa -I _tmp.cram -O _tmp.sam
sq1     0       c1      1       0       10M     *       0       0       AACCGCGGTT      **********
sQ1     0       c1      1       0       10M     *       0       0       AACCGCGGTT      *
SQ1     0       c1      1       0       10M     *       0       0       *       *       MD:Z:10 NM:i:0
sq2     0       c1      1       0       4M1D5M  *       0       0       AACCCGGTT       *********
sQ2     0       c1      1       0       4M1D5M  *       0       0       AACCCGGTT       *
SQ2     0       c1      1       0       4M1D5M  *       0       0       *       *       MD:Z:4^G5  NM:i:1
sq3     4       c1      1       0       *       *       0       0       AACCCGGTT       *********
sQ3     4       c1      1       0       *       *       0       0       AACCCGGTT       *
SQ3     4       c1      1       0       *       *       0       0       *       *
```

**Comment by jkbonfield** [comment]
Oops, SD is score digest, so it's the quality values not the sequences that claim to differ.

**Comment by jkbonfield** [comment]
Checking how my code computes this, the above cigar/seq/quality strings are:

10M AACCGCGGTT **********  => checksum ********** (actually "\t" x10 as "*" is Q9)
10M AACCGCGGTT * => checksum <255>x10 (unknown quality is encoded as 255 in BAM format)
10M * * => checksum "" (empty string)

Basically it's checksumming what would be in the BAM file.  For seq "AAC" and quality "*" BAM will store both as length 3 with the quality being <255><255><255>, but for seq * quality * BAM will store both as 0 length.

We need to be explicit in the documentation on this, and agree (obviously!).  How is Cramtools counting quality bases for computing the checksums?

**Comment by vadimzalunin** [comment]
*I THINK* I fixed it: 8adb5f3df344a906a6fdd5f0c611959af2c857eb

**Comment by jkbonfield** [comment]
Seems to work, so closing it.
